### PR TITLE
Support Python 3.9

### PR DESCRIPTION
**Pull request by goodoldneon**
## Description
Add Python 3.9 support. This required 2 changes:
- Use `typing.Optional` and `typing.Union` instead of the `|` operator ([PEP 604](https://peps.python.org/pep-0604))
- Use `typing_extensions.TypeGuard` instead of `typing.TypeGuard` ([PEP 647](https://peps.python.org/pep-0647))

Additionally:
- Use `3.9` in CI. Stopped using `3.10` and `3.11` in CI since it's overkill: it seems OK to use only the lowest and highest Python versions we support
- Update minimum `typing_extensions` to `4.8.0`, since this is what `fastapi==0.104.0` requires

## Issue
#61

## Review
This PR has a lot of changes but they're very low risk since they're almost entirely type annotations. We have a thorough test suite, so these changes should be OK as long as everything passes

### CODE REVIEW

**Comment by jpwilliams** [comment]
@goodoldneon Can user code can still use the newer features like the `|` operator even though the types are created using `typing.Union`?

Also we should update the badge on the README to show `3.9` and maybe supported version ranges of the frameworks too?

**Comment by goodoldneon** [comment]
> Can user code can still use the newer features like the | operator even though the types are created using typing.Union?

Yes, type checkers treat `typing.Union` the same way as `|`, so users are free to continue using `|` in their code.

> Also we should update the badge on the README to show 3.9 and maybe supported version ranges of the frameworks too?

The badge will automatically update when we publish next. I added framework versions to the readme

File: README.md [hunk 1]
```diff
@@ -27,10 +27,12 @@
...
-- Django
-- FastAPI
-- Flask
-- Tornado
+- Django (`>=4.2`)
+- FastAPI (`>=0.100.0`)
+- Flask (`>=2.3.0`)
+- Tornado (`>=6.3`)
+
+Python 3.9 is the minimum version we support.
```
Comment (djfarrelly) [inline README.md:35]: Nice addition w/ version compat!

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -35,7 +35,7 @@ jobs:
     runs-on: "ubuntu-latest"
     strategy:
       matrix:
-        python-version: ["3.10", "3.11", "3.12"]
+        python-version: ["3.9", "3.12"]
```
Comment (djfarrelly) [inline .github/workflows/ci.yml:38]: Is there any value in running against 3.10 and 3.11 here as well or are you assuming that there aren't major changes between those versions that affect the SDK so it isn't necessary to run the suite against them?

**Review by djfarrelly — APPROVED** [review]
3.9 support looks great!

File: .github/workflows/ci.yml (hunk 2 above)
Comment (goodoldneon) [reply .github/workflows/ci.yml:38]: The benefit seems really small compared to the downside of extra jobs making CI slower

**Pull request merged**